metricflow/model/validations/identifiers.py

```diff
@@ -119,3 +123,90 @@ def validate_model(model: UserConfiguredModel) -> List[ValidationIssue]:  # noqa
+    def validate_model(model: UserConfiguredModel) -> List[ValidationIssueType]:  # noqa: D
+        issues: List[ValidationIssueType] = []
+        # The expected sub identifiers that an identifier has.
+        identifier_to_sub_identifiers: OrderedDict[str, Dict[str, Tuple[str, ...]]] = OrderedDict()
+        # The identifiers that have different sub-identifiers in different data sources.
+        identifiers_with_inconsistent_sub_identifiers = []
+
+        identifier_contexts = []
+        issues = []
+        for data_source in model.data_sources or []:
+            sub_identifier_contexts = IdentifierConsistencyRule._get_sub_identifier_context(data_source)
+            for sub_identifier_context in sub_identifier_contexts:
+                identifier_contexts.append(sub_identifier_context)
+
+                identifier_reference = sub_identifier_context.identifier_reference
+                if identifier_reference.element_name not in identifier_to_sub_identifiers:
+                    identifier_to_sub_identifiers[identifier_reference.element_name] = {
+                        data_source.name: sub_identifier_context.sub_identifier_names
+                    }
+
+                elif (
+                    sub_identifier_context.sub_identifier_names
+                    not in identifier_to_sub_identifiers[identifier_reference.element_name].values()
+                ):
+                    identifier_to_sub_identifiers[identifier_reference.element_name][
+                        data_source.name
+                    ] = sub_identifier_context.sub_identifier_names
+                    identifiers_with_inconsistent_sub_identifiers.append(identifier_reference)
+
+        for identifier_reference in identifiers_with_inconsistent_sub_identifiers:
+            for identifier_context in identifier_contexts:
+                if identifier_context.identifier_reference.element_name == identifier_reference.element_name:
+                    issues.append(
+                        ValidationWarning(
+                            model_object_reference=ValidationIssue.make_object_reference(
+                                data_source_name=identifier_context.data_source_name,
+                                identifier_name=identifier_context.identifier_reference.element_name,
+                            ),
+                            message=(
+                                f"Identifier '{identifier_reference.element_name}' does not have consistent sub-identifiers "
+                                f"throughout the model: {identifier_to_sub_identifiers[identifier_reference.element_name]}. "
+                                f"Specifically, in data source '{identifier_context.data_source_name}', "
+                                f"it has conflicting sub-identifiers {list(identifier_context.sub_identifier_names)}."
+                            ),
+                        )
+                    )
```

Comment:
This is very hard to understand because the assertion and message construction logic is now all tangled up with the loop iteration mechanics we're using to parse the data source model information. I think it'd be cleaner if we instead did something like:

```
Build mapping of identifiers to sub-identifiers, with a set of associated data sources
for every element in mapping:
    check for mismatched sub-identifiers
    add errors if necessary
```
This has the added benefit of allowing us to aggregate our error messages on a per-identifier basis. I'm thinking something like this, but actually, you know, working:
```suggestion
        # The expected sub identifiers that an identifier has.
        identifier_to_sub_identifiers: OrderedDict[str, DefaultDict[Tuple[str, ...], Set[str]]] = OrderedDict()

        for data_source in model.data_sources or []:
            sub_identifier_contexts = IdentifierConsistencyRule._get_sub_identifier_context(data_source)
            for sub_identifier_context in sub_identifier_contexts:
                identifier_reference = sub_identifier_context.identifier_reference
                if identifier_reference.element_name not in identifier_to_sub_identifiers:
                    sub_identifiers = collections.defaultdict(set)
                    sub_identifiers[sub_identifier_context.identifier_names].add(data_source.name)
                    identifier_to_sub_identifiers[identifier_reference.element_name] = sub_identifiers
                else:
                    identifier_to_sub_identifiers[identifier_reference.element_name][
                        sub_identifier_context.sub_identifier_names
                    ].add(data_source.name)

        for identifier_name, sub_identifiers in identifier_to_sub_identifiers.items():
            if len(sub_identifiers) < 2:
                continue
                
            first_data_source_name = sub_identifiers[0][0]
            issues.append(
                ValidationWarning(
                    model_object_reference=ValidationIssue.make_object_reference(
                        data_source_name=first_data_source_name,
                        identifier_name=identifier_name,
                    ),
                    message=(
                        f"Identifier '{identifier_name}' does not have consistent sub-identifiers "
                        f"throughout the model. We found {len(sub_identifiers)} sets of sub-identifiers, but there "
                        f"should be no more than 1! Here are all sub-identifier sets with associated data sources: "
                        f"{sub_identifiers}"
                    ),
                )
            )
```

Reply:
Worth noting, these loops can each be a separate function with a nice name as well, which would aid in readability. Might be worth doing the first one, so `validate_model` becomes:

```
identifier_to_sub_identifiers = IdentifierConsistencyRule._build_sub_identifier_mapping(model.data_sources or [])

for identifier_name, sub_identifiers in identifier_to_sub_identifiers.items():
    # etc.

return issues
```
You can do both of course but the second loop is both pretty straightforward and is doing the thing the so I don't think it buys you much, and anyway that's actually doing the validation work.

This probably isn't necessary, but it makes that first gnarly for loop a little more accessible to a reader.

Reply:
Oh wait, we are working way too hard here. The data source names are in the sub-identifier contexts so we don't need to collect them again along the way, so we can do something more like aggregate -> filter -> convert.

Note - it'd be nicer to do some more work in the last block to print one line per sub-identifier list value with collections of data sources, instead of a list of all non-unique values, but I'm tired of writing code in GitHub comments.

```suggestion
        # build collection of sub-identifier contexts, keyed by identifier name
        identifier_to_sub_identifier_contexts: DefaultDict[str, List[SubIdentifierContext]] = collections.defaultdict(list)
        all_contexts: List[SubIdentifierContext]] = more_itertools.flatten([
            IdentifierConsistencyRule._get_sub_identifier_context(data_source)
            for data_source in model.data_sources
        ])
        for context in all_contexts:
            identifier_to_sub_identifier_contexts[
                context.identifier_reference.element_name
            ].append(context)
            
        # Filter out anything that has fewer than 2 distinct sub-identifier sets
        invalid_sub_identifier_configurations = dict(filter(
            lambda item: len(
                set([context.sub_identifier_names for context in item[1]])
            ) >= 2,
            identifier_to_sub_identifier_contexts.items()
        ))

        # convert each invalid identifier configuration into a validation warning 
        for identifier_name, sub_identifier_contexts in invalid_sub_identifier_configurations.items():
            issues.append(
                ValidationWarning(
                    model_object_reference=ValidationIssue.make_object_reference(
                        data_source_name=sub_identifier_contexts[0].data_source_name,
                        identifier_name=identifier_name,
                    ),
                    message=(
                        f"Identifier '{identifier_name}' does not have consistent sub-identifiers "
                        f"throughout the model: {list(sorted(sub_identifier_contexts, key=lambda x: x.sub_identifier_names)}"
                    ),
                )
            )
```

Reply:
Added.